keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,309 @@
+
+- `component` (used by [kubeadm][kubeadm-labels])
+- `tier` (used by [kubeadm][kubeadm-labels])
+- `k8s-app` (common on [addons][addons-k8s-app])
+
+The `k8s-app` label is used to match controllers for system components, and therefore should be
+explicitly disallowed.
+
+It is not clear how the `tier` and `component` labels are consumed, but I recommend uses be migrated
+to `unrestricted.node.kubernetes.io/component` and `unrestricted.node.kubernetes.io/tier`, rather
+than special casing the existing labels.
+
+[kubeadm-labels]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cmd/kubeadm/app/util/staticpod/utils.go#L60
+[addons-k8s-app]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cluster/addons/kube-proxy/kube-proxy-ds.yaml#L23
+
+### Annotation Restrictions
+
+The Kubelet will be prevented from updating pod annotations or creating mirror pods with
+annotations, except for whitelisted keys:
+
+1. Any annotations starting with `unrestricted.node.kubernetes.io/` are allowed.
+2. Annotations the Kubelet currently uses are allowed:
+- `ConfigMirrorAnnotationKey = "kubernetes.io/config.mirror"`
+- `ConfigHashAnnotationKey = "kubernetes.io/config.hash"`
+- `ConfigFirstSeenAnnotationKey = "kubernetes.io/config.seen"`
+- `ConfigSourceAnnotationKey = "kubernetes.io/config.source"`
+3. Well-known annotations that may be used on static pods are allowed:
+- `PodPresetOptOutAnnotationKey = "podpreset.admission.kubernetes.io/exclude"
+- `SeccompPodAnnotationKey = "seccomp.security.alpha.kubernetes.io/pod"`
+- `SeccompContainerAnnotationKeyPrefix = "container.seccomp.security.alpha.kubernetes.io/"` (prefix
+  match)
+- `ContainerAnnotationKeyPrefix = "container.apparmor.security.beta.kubernetes.io/"` (prefix match)
+- `PreferAvoidPodsAnnotationKey = "scheduler.alpha.kubernetes.io/preferAvoidPods"`
+- `BootstrapCheckpointAnnotationKey = "node.kubernetes.io/bootstrap-checkpoint"`
+
+Note that annotation restrictions are applying the least-privilege principal, but are not directly
+mitigating a known threat.
+
+### OwnerReferences
+
+OwnerReferences cannot be updated through the `pod/status` subresource, but they can be set on
+mirror pods. With the new restrictions, mirror pods are only allowed a single owner reference (or
+none), and it must refer to the node:
+
+```go
+ metav1.OwnerReference{
+  APIVersion: "v1"
+  Kind: "Node"
+  Name: node.Name
+  UID:  node.UID
+  Controller: nil // or false
+  BlockOwnerDeletion: nil // all values allowed
+}
+```
+
+### Risks and Mitigations
+
+Some Kubernetes setups depend on statically serving services today. Applying these mitigations will
+likely break these clusters. There is no way to apply these changes in a fully backwards compatible
+way, so instead we will rely on a staged rollout through the `NodeRestrictionPods` feature gate, and
+call out the actions required.
```

Comment:
it seems like we could do more here to ensure an upgraded cluster would continue working, especially around labels (like doing a scan on startup for static pods with out-of-bounds labels and continuing to allow those specific labels in addition to the proposed prefixes).

I'd also expect to roll this out in the API server and in the kubelet in separate releases (the same way we tightened which node labels the kubelet will apply 3 releases before the API server starts enforcing)

Reply:
> like doing a scan on startup for static pods with out-of-bounds labels and continuing to allow those specific labels in addition to the proposed prefixes

How would that work without creating a backdoor to get around the restrictions? I suppose that it could check for existing mirror pods, and auto-whitelist the labels? I would want to provide a way for cluster admins to see the implicit whitelist and override it.

> I'd also expect to roll this out in the API server and in the kubelet in separate releases

The proposal as it stands doesn't include any Kubelet changes. Did you have something in mind? I suppose the Kubelet could filter the disallowed fields before they are rejected.

Reply:
> I suppose that it could check for existing mirror pods, and auto-whitelist the labels?

That's what I was envisioning, though I hadn't spent much time thinking of alternatives. I want to make sure we don't make a change in one release in the API server that takes a working cluster and breaks it.

> I would want to provide a way for cluster admins to see the implicit whitelist and override it.

Yeah, that's back to the convention vs configuration quandry. The more invasive the hard-coded policy, the more admins want the ability to configure it.

>> I'd also expect to roll this out in the API server and in the kubelet in separate releases
>
> The proposal as it stands doesn't include any Kubelet changes. Did you have something in mind? I suppose the Kubelet could filter the disallowed fields before they are rejected.

In general, it doesn't seem good to hardcode policy in the API server that rejects operations the kubelet will happily perform. By making the kubelet aware of the limits, and making it error (not silently drop offending data) in release _n_, then waiting until _n+3_ to restrict the behavior in the API server, you ensure that you are not breaking a working node configuration.

Reply:
If kubelet auto-removes labels, and adds them to an annotation, it would be possible to write a controller that reverses the operation. This could be implemented as a pod that takes a pre-configured whitelist of allowed mirror-pod labels, and watches pods.

On seeing a mirror-pod addition:

1. Check for the `kubernetes.io/config.removed-labels` annotation
2. If it exists, deserialize the json to a label map
3. Check removed label keys against whitelist
4. If label is whitelisted, update pod to add the label back.

We could ship something like this alongside this feature, but it seems like overkill...